### PR TITLE
feat: [CDS-39032]: Azure web app common

### PR DESCRIPTION
**Pull request by Tamaraius**
### Summary

Figma: https://www.figma.com/file/aKWbWYzou4B8YXcnifcHVj/Azure-Web-App-CDNG?node-id=69%3A105588

#### Screenshots

https://user-images.githubusercontent.com/98325173/178743645-d1ff0579-3d40-4247-ac02-0b7aedaf0bb1.mov

image

image

https://user-images.githubusercontent.com/98325173/178744442-84dbc700-7cdd-4918-9e58-d0b2a19494c0.mov

https://user-images.githubusercontent.com/98325173/178744862-2abc4954-592d-4c7f-a21b-ccd49f508751.mov

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>
<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by YounixIO** [comment]
retrigger titlecheck

File: src/modules/35-connectors/components/ConnectorReferenceField/FormMultiTypeConnectorField.tsx [hunk 1]
```diff
@@ -78,6 +78,7 @@ export interface MultiTypeConnectorFieldProps extends Omit<ConnectorReferenceFie
   multitypeInputValue?: MultiTypeInputType
   connectorLabelClass?: string
   onLoadingFinish?: () => void
+  setConnector?: any
```
Comment (bhavya-sinha) [inline src/modules/35-connectors/components/ConnectorReferenceField/FormMultiTypeConnectorField.tsx:81]: Why is this any?

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactHelper.ts [hunk 2]
```diff
@@ -35,6 +43,16 @@ export const isAdditionAllowed = (deploymentType: ServiceDefinition['type'], isR
   )
 }
 
+export const isAdditionAllowed = (deploymentType: ServiceDefinition['type'], isReadOnly: boolean): boolean => {
+  return (
+    !isReadOnly &&
+    (deploymentType === ServiceDeploymentType.Kubernetes ||
+      deploymentType === ServiceDeploymentType.NativeHelm ||
+      deploymentType === ServiceDeploymentType.ServerlessAwsLambda ||
+      deploymentType === ServiceDeploymentType.AzureWebApp)
```
Comment (bhavya-sinha) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactHelper.ts:52]: No need to add additional function. Just add `deploymentType === ServiceDeploymentType.AzureWebApp` in `isAdditionAllowed`

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactHelper.ts [hunk 3]
```diff
@@ -22,11 +22,19 @@ export enum ModalViewFor {
...
+    deploymentType === ServiceDeploymentType.AzureWebApp
+  )
+}
+
+export const isAllowedCustomArtifactDeploymentTypes = (deploymentType: ServiceDefinition['type']): boolean => {
   return deploymentType === ServiceDeploymentType.Kubernetes || deploymentType === ServiceDeploymentType.NativeHelm
 }
 
-export const isAdditionAllowed = (deploymentType: ServiceDefinition['type'], isReadOnly: boolean): boolean => {
+export const isSidecarAllowed = (deploymentType: ServiceDefinition['type'], isReadOnly: boolean): boolean => {
```
Comment (bhavya-sinha) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactHelper.ts:37]: Remove this func

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactListView/ArtifactListView.tsx [hunk 4]
```diff
@@ -269,7 +270,7 @@ function ArtifactListView({
             text={getString('pipelineSteps.serviceTab.artifactList.addPrimary')}
           />
         )}
-        {isAdditionAllowed && (
+        {isAdditionAllowed && isSidecarAllowed && (
```
Comment (bhavya-sinha) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactListView/ArtifactListView.tsx:273]: Update this condition

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactsSelection.tsx [hunk 5]
```diff
@@ -135,15 +137,15 @@ export default function ArtifactsSelection({
...
+      isAllowedCustomArtifactDeploymentTypes(deploymentType)
     ) {
       allowedArtifactTypes[deploymentType].push(ENABLED_ARTIFACT_TYPES.CustomArtifact)
     }
 
     if (
       NG_AZURE &&
       !allowedArtifactTypes[deploymentType]?.includes(ENABLED_ARTIFACT_TYPES.Acr) &&
-      isAllowedArtifactDeploymentTypes(deploymentType)
+      isAllowedACRArtifactDeploymentTypes(deploymentType)
```
Comment (bhavya-sinha) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactsSelection.tsx:148]: Please use the old function, as this is not just ACR specific

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactHelper.ts [hunk 6]
```diff
@@ -22,11 +22,19 @@ export enum ModalViewFor {
   SIDECAR = 2
 }
 
-export const isAllowedArtifactDeploymentTypes = (deploymentType: ServiceDefinition['type']): boolean => {
+export const isAllowedACRArtifactDeploymentTypes = (deploymentType: ServiceDefinition['type']): boolean => {
```
Comment (bhavya-sinha) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactHelper.ts:25]: This function is not needed

File: src/modules/70-pipeline/components/ManifestSelection/ManifestInterface.ts [hunk 7]
```diff
@@ -7,14 +7,14 @@
 
 import type { MultiTypeInputType, SelectOption } from '@wings-software/uicore'
 import type { ConnectorSelectedValue } from '@connectors/components/ConnectorReferenceField/ConnectorReferenceField'
+import type { PipelineInfoConfig } from 'services/pipeline-ng'
 import type {
-  ConnectorConfigDTO,
   ManifestConfig,
+  ConnectorConfigDTO,
```
Comment (bhavya-sinha) [inline src/modules/70-pipeline/components/ManifestSelection/ManifestInterface.ts:13]: Why is this added back?

Comment (bhavya-sinha) [reply src/modules/70-pipeline/components/ManifestSelection/ManifestInterface.ts:13]: Revert this file changes

**Review by bhavya-sinha — CHANGES_REQUESTED** [review]

File: src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecEditable.tsx [hunk 8]
```diff
@@ -0,0 +1,850 @@
...
+    }
+    return () =>
+      unSubscribeForm({
+        tab: DeployTabs.INFRASTRUCTURE,
+        form: formikRef as React.MutableRefObject<FormikProps<unknown> | null>
+      })
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
+  const isSubscriptionDisabled = (connectorValue: AzureFieldTypes): boolean => {
+    if (getValue(connectorValue)?.length) {
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecEditable.tsx:294]: Can be written like return `!getValue(connectorValue)?.length`

File: src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecEditable.tsx [hunk 9]
```diff
@@ -0,0 +1,850 @@
...
+      return false
+    }
+    return true
+  }
+  const isResourceGroupDisabled = (subscriptionValue: AzureFieldTypes, connectorValue: AzureFieldTypes): boolean => {
+    if (getValue(subscriptionValue)?.length && !isSubscriptionDisabled(connectorValue)) {
+      return false
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecEditable.tsx:301]: same as above

File: src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecEditable.tsx [hunk 10]
```diff
@@ -0,0 +1,850 @@
...
+    }
+
+    return currentValues
+  }
+
+  const { subscribeForm, unSubscribeForm } = React.useContext(StageErrorContext)
+
+  React.useEffect(() => {
+    subscribeForm({
+      tab: DeployTabs.INFRASTRUCTURE,
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecEditable.tsx:210]: Why is this `subscribeForm` needed?

File: src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecEditable.tsx [hunk 11]
```diff
@@ -0,0 +1,850 @@
...
+  }
+
+  return (
+    <Layout.Vertical spacing="medium">
+      <Formik<AzureWebAppInfrastructureUI>
+        formName="azureWebAppInfra"
+        initialValues={getInitialValues()}
+        validate={value => {
+          const data: Partial<AzureWebAppInfrastructure> = {
+            connectorRef: undefined,
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecEditable.tsx:334]: Extract this to separate function. It's not readable

File: src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecEditable.tsx [hunk 12]
```diff
@@ -0,0 +1,850 @@
...
+                  projectIdentifier={projectIdentifier}
+                  orgIdentifier={orgIdentifier}
+                  width={450}
+                  connectorLabelClass={css.connectorRef}
+                  enableConfigureOptions={false}
+                  style={{ marginBottom: 'var(--spacing-large)' }}
+                  type={Connectors.AZURE}
+                  gitScope={{ repo: repoIdentifier || '', branch, getDefaultFromOtherRepo: true }}
+                  onChange={type => {
+                    if (type !== MultiTypeInputType.FIXED) {
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecEditable.tsx:383]: Extract this out

File: src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecEditable.tsx [hunk 13]
```diff
@@ -0,0 +1,850 @@
...
+                  selectItems={subscriptions}
+                  disabled={isSubscriptionDisabled(formik.values?.connectorRef) || readonly}
+                  placeholder={
+                    loadingSubscriptions
+                      ? /* istanbul ignore next */ getString('loading')
+                      : getString('cd.steps.azureInfraStep.subscriptionPlaceholder')
+                  }
+                  multiTypeInputProps={{
+                    onChange: /* istanbul ignore next */ () => {
+                      getMultiTypeFromValue(formik.values?.resourceGroup) !== MultiTypeInputType.RUNTIME &&
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecEditable.tsx:433]: Extract this

File: src/modules/70-pipeline/components/ManifestSelection/ManifestInterface.ts (hunk 7 above)
Comment (YounixIO) [reply src/modules/70-pipeline/components/ManifestSelection/ManifestInterface.ts:13]: This is the current change on develop branch.

Comment (YounixIO) [reply src/modules/70-pipeline/components/ManifestSelection/ManifestInterface.ts:13]: https://github.com/harness/harness-core-ui/blob/develop/src/modules/70-pipeline/components/ManifestSelection/ManifestInterface.ts

Comment (YounixIO) [reply src/modules/70-pipeline/components/ManifestSelection/ManifestInterface.ts:13]: Screenshot 2022-07-13 at 5 26 34 PM

File: src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecInputForm.tsx [hunk 14]
```diff
@@ -0,0 +1,716 @@
...
+        <div className={cx(stepCss.formGroup, stepCss.md, css.inputWrapper)}>
+          <FormInput.MultiTypeInput
+            name={`${path}.deploymentSlot`}
+            tooltipProps={{
+              dataTooltipId: 'azureInfraDeploymentSlot'
+            }}
+            disabled={
+              !(
+                initialValues.connectorRef?.length &&
+                initialValues.connectorRef !== '<+input>' &&
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/AzureWebAppInfrastructureStep/AzureWebAppInfrastructureSpecInputForm.tsx:576]: This is not readable. Please move this to function

File: src/modules/70-pipeline/utils/stageHelpers.ts [hunk 15]
```diff
@@ -434,6 +439,9 @@ export const getStepTypeByDeploymentType = (deploymentType: string): StepType =>
   if (isServerlessDeploymentType(deploymentType)) {
     return StepType.ServerlessAwsLambda
   }
+  if (deploymentType === ServiceDeploymentType.AzureWebApp) {
+    return StepType.AzureWebAppServiceSpec
```
Comment (bhavya-sinha) [inline src/modules/70-pipeline/utils/stageHelpers.ts:443]: Why is this done?

Comment (YounixIO) [reply src/modules/70-pipeline/utils/stageHelpers.ts:443]: We wanted to have all the AzureWebApp service related changes into a separate service file as Kuberentes service file was getting bloated with if else block. Hence the change.

File: src/modules/75-cd/components/PipelineSteps/AzureWebAppServiceSpec/AzureWebAppServiceSpec.tsx [hunk 16]
```diff
@@ -0,0 +1,509 @@
...
+          readonly={readonly}
+          allowableTypes={allowableTypes}
+        />
+      )
+    }
+
+    if (stepViewType === StepViewType.InputSet || stepViewType === StepViewType.DeploymentForm) {
+      return (
+        <AzureWebAppServiceSpecInputSetForm
+          {...(customStepProps as AzureWebAppServiceSpecFormProps)}
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/AzureWebAppServiceSpec/AzureWebAppServiceSpec.tsx:483]: ` <KubernetesServiceSpecInputSetMode` Use this

File: src/modules/70-pipeline/components/ManifestSelection/ManifestInterface.ts (hunk 7 above)
Comment (bhavya-sinha) [reply src/modules/70-pipeline/components/ManifestSelection/ManifestInterface.ts:13]: Looks like the branch is not rebased with develop, else the diff should not be there

**Review by bhavya-sinha — DISMISSED** [review]

**Review by kapilkumawat86 — APPROVED** [review]

**Review by kapilkumawat86 — APPROVED** [review]

**Review by YounixIO — APPROVED** [review]

**Pull request merged**